mlos_core/mlos_core/optimizers/observations.py

```diff
@@ -0,0 +1,361 @@
+#
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+#
+"""Contains classes for observations and suggestions."""
+
+from __future__ import annotations
+
+from typing import Any, Iterator, List, Optional
+
+import pandas as pd
+from ConfigSpace import Configuration, ConfigurationSpace
+
+
+def compare_optional_series(left: Optional[pd.Series], right: Optional[pd.Series]) -> bool:
+    """
+    Compare Series that may also be None.
+
+    Parameters
+    ----------
+    left : Optional[pd.Series]
+        The left Series to compare
+    right : Optional[pd.Series]
+        The right Series to compare
+
+    Returns
+    -------
+    bool
+        Compare the equality of two Optional[pd.Series] objects
+    """
+    if type(left) != type(right):
+        return False
+    elif left is not None and right is not None:
+        if not left.equals(right):
+            return False
+    return True
+
+
+def compare_optional_dataframe(
+    left: Optional[pd.DataFrame], right: Optional[pd.DataFrame]
+) -> bool:
+    """
+    Compare DataFrames that may also be None.
+
+    Parameters
+    ----------
+    left : Optional[pd.DataFrame]
+        The left DataFrame to compare
+    right : Optional[pd.DataFrame]
+        The right DataFrame to compare
+
+    Returns
+    -------
+    bool
+        Compare the equality of two Optional[pd.DataFrame] objects
+    """
+    if type(left) != type(right):
+        return False
+    elif left is not None and right is not None:
+        if not left.equals(right):
+            return False
+    return True
+
+
+class Observation:
+    """
+    A single observation of a configuration.
+
+    Attributes
+    ----------
+    config : pd.Series
+        The configuration observed.
+    score : pd.Series
+        The score metrics observed.
+    context : Optional[pd.Series]
+        The context in which the configuration was evaluated.
+        Not Yet Implemented.
+    metadata: Optional[pd.Series]
+        The metadata in which the configuration was evaluated
+        Not Yet Implemented.
+    """
+
+    def __init__(
+        self,
+        *,
+        config: pd.Series = pd.Series(),
+        score: pd.Series = pd.Series(),
+        context: Optional[pd.Series] = None,
+        metadata: Optional[pd.Series] = None,
+    ):
+        self.config = config
+        self.score = score
+        self.context = context
+        self.metadata = metadata
+
+    def to_suggestion(self) -> Suggestion:
```

Comment:
```suggestion
    def to_suggestion(self) -> "Suggestion":
```

Reply:
That should let you work around the order of class definitions.

Reply:
You can add something like the following at the top as well (in case of splitting things out):

```python
if TYPE_CHECKING:
    from mlos_core.data_types.suggestion import Suggestion
```